[c10d] add ncclGetLastError to NCCL pg

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #83724

This PR add ncclGetLastError API to the nccl pg, to provide better error
reporting out of nccl failures directly, instead of guessing on random
reasons

Differential Revision: [D39161199](https://our.internmc.facebook.com/intern/diff/D39161199)

wanchaol committed:
[c10d] add ncclGetLastError to NCCL pg

This PR add ncclGetLastError API to the nccl pg, to provide better error
reporting out of nccl failures directly, instead of guessing on random
reasons

[ghstack-poisoned]

kwen2501 reviewed (approved):
LGTM!
nit: 
(i) Shall we print NCCL errors on a new line so that they stand out?
(ii) We can use a string variable (e.g. `interpret`) to hold the error interpretation. Then, at the end, we only need to do once:
`return (interpret + "\n Last error:\n" + err).c_str();`

H-Huang commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -21,22 +21,23 @@ const inline char* getNcclErrorDetailStr(ncclResult_t error, c10::optional<std::
   if (processGroupFailureReason != c10::nullopt) {
     return (*processGroupFailureReason).c_str();
   }
+  std::string err = std::string(ncclGetLastError(NULL));
```
Do we need to check the NCCL version before calling this API? https://docs.nvidia.com/deeplearning/nccl/user-guide/docs/api/comms.html#ncclgetversion

wanchaol replied:
Given that our nccl version both in OSS and internal already have this API, I think we should be fine since this code will ship together with the newer nccl version?

H-Huang replied:
It does look like jobs in https://www.internalfb.com/diff/D39161199 are failing due to this new function call. I noticed that we can use a #if directive to describe using this feature. I will add example, in a separate comment.

H-Huang reviewed (commented):
Thanks for adding, this will be super helpful!

I think we should also import this as a diff to do a sanity check internally. As I understand it the latest NCCL version is available (12.3.4), but we don't want to break anyone that is using an older version.

rohan-varma reviewed (commented):
Thanks, super excited to see this in PT! Could we potentially add a unittest or even some print outs of what the error looks like?

wanchaol commented:
> Could we potentially add a unittest or even some print outs of what the error looks like?

The problem with this change is that it's hard to trigger the nccl failures, especially all types of failures in this PR. Do you have suggestions how we could better test those failures? I looked into our current testing code and it seems that we only test the time out issues, not any other failures.

wanchaol committed:
Update on "[c10d] add ncclGetLastError to NCCL pg"


This PR add ncclGetLastError API to the nccl pg, to provide better error
reporting out of nccl failures directly, instead of guessing on random
reasons

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39161199).

H-Huang commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -21,22 +21,29 @@ const inline char* getNcclErrorDetailStr(ncclResult_t error, c10::optional<std::
     return (*processGroupFailureReason).c_str();
   }
+  std::string interpret = nullptr;
+  std::string err = std::string(ncclGetLastError(NULL));
```
We can wrap the `ncclGetLastError()` in an if directive

Directive definition example: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/NCCLUtils.hpp#L44-L49

Directive usage example:
https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/NCCLUtils.hpp#L115-L122

wanchaol committed:
Update on "[c10d] add ncclGetLastError to NCCL pg"


This PR add ncclGetLastError API to the nccl pg, to provide better error
reporting out of nccl failures directly, instead of guessing on random
reasons

Differential Revision: [D39161199](https://our.internmc.facebook.com/intern/diff/D39161199)

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39161199).

H-Huang commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -21,24 +21,43 @@ const inline char* getNcclErrorDetailStr(ncclResult_t error, c10::optional<std::
+      interpret = "Unknown NCCL error!";
   }
-  return "Unknown NCCL error";
+  return (interpret + err).c_str();
```
Thanks for the directive changes! looks great.

I just saw the new errors in [D39161199](https://www.internalfb.com/diff/D39161199) and I think it is because we are return char* for strings allocated locally. It worked before because we were only returning string literals (https://stackoverflow.com/questions/2371668/is-returning-a-string-literal-address-from-a-function-safe-and-portable/2371698#2371698)

We should just change `getNcclErrorDetailStr()` to return `const std::string`, I don't think there is a particular reason it is returning `char*`

H-Huang commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -21,24 +21,43 @@ const inline char* getNcclErrorDetailStr(ncclResult_t error, c10::optional<std::
 }
 } // namespace
+
+// GetLastError is enabled only for NCCL versions 2.13+
```
nit:
```suggestion
// GetLastError() is enabled only for NCCL versions 2.13+
```

wanchaol committed:
Update on "[c10d] add ncclGetLastError to NCCL pg"


This PR add ncclGetLastError API to the nccl pg, to provide better error
reporting out of nccl failures directly, instead of guessing on random
reasons

Differential Revision: [D39161199](https://our.internmc.facebook.com/intern/diff/D39161199)

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39161199).

wanchaol committed:
Update on "[c10d] add ncclGetLastError to NCCL pg"


This PR add ncclGetLastError API to the nccl pg, to provide better error
reporting out of nccl failures directly, instead of guessing on random
reasons

Differential Revision: [D39161199](https://our.internmc.facebook.com/intern/diff/D39161199)

[ghstack-poisoned]

wanchaol commented:
@wanchaol has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39161199).

H-Huang reviewed (approved):
Looks great! Thanks for making these changes, they will be super helpful in providing better error messages for NCCL!

wanchaol commented:
@pytorchbot merge
